src/lib.rs

```diff
@@ -416,12 +442,19 @@ impl<T: Translation> Mapping<T> {
     /// and modifying those would violate architectural break-before-make (BBM) requirements.
     pub fn modify_range<F>(&mut self, range: &MemoryRegion, f: &F) -> Result<(), MapError>
     where
-        F: Fn(&MemoryRegion, &mut Descriptor, usize) -> Result<(), ()> + ?Sized,
+        F: Fn(&MemoryRegion, &mut Descriptor, usize) -> Result<bool, ()> + ?Sized,
     {
         if self.active() {
             self.check_range_bbm(range, f)?;
         }
-        self.root.modify_range(range, f)?;
+        if self.root.modify_range(range, f, self.active())? && self.active() {
+            // SAFETY: Barriers have no side effects that are observeable by the program
+            #[cfg(target_arch = "aarch64")]
+            unsafe {
+                // Wait for the TLBI to complete
+                asm!("dsb ish", "isb", options(preserves_flags));
```

Comment:
We call this `dsb ish` `isb` in a few places now, so it would be worth factoring out to a safe helper function.

I think you can add the `nostack` option too.

I also noticed that in some places in the code we use `dsb ishst`, some places `dsb nsh`, and here `dsh ish`. Is this correct?

Reply:
Yes I'll factor out `dsb ish` + `isb` - those are the ones that are needed to ensure that all CPUs observe the updated permissions on live mappings.

`dsb ishst` is needed to complete any stores before enabling the translation, so that the page table walker is guaranteed not to see stale data. `dsb nsh` is sufficient to complete TLB maintenance on the local CPU when deactivating the mapping, this is also correct afaict.